### PR TITLE
New version of keychain dump

### PR DESCRIPTION
**Pull request by tghosth**
As discussed in #35 Updated version of keychain_dump which uses the binary from here:
https://github.com/dmayer/idb/tree/master/lib/utils/keychain_dump

This works much better than the keychaindumper which was previously used which was missing items.

Added at the same time a new method to remote_operations which allows a file to be read using a grep filter.

### CODE REVIEW

File: needle/modules/storage/data/keychain_dump.py [hunk 1]
```diff
@@ -32,23 +35,54 @@ def module_pre(self):
     # ==================================================================================================================
     def module_run(self):
         # Composing the command string
-        cmd = '{}'.format(self.device.DEVICE_TOOLS['KEYCHAINDUMPER'])
+        cmd = '{}'.format(self.device.DEVICE_TOOLS['KEYCHAIN_DUMP'])
```
Comment (marco-lancini) [inline needle/modules/storage/data/keychain_dump.py]: this creates plist files in the user's home directory. Is there any way to force a specified output directory?

**Comment by marco-lancini** [comment]
Had to revert it cause it is not ready to be pushed out

File: needle/modules/storage/data/keychain_dump.py [hunk 2]
```diff
@@ -32,23 +35,54 @@ def module_pre(self):
     # ==================================================================================================================
     def module_run(self):
         # Composing the command string
-        cmd = '{}'.format(self.device.DEVICE_TOOLS['KEYCHAINDUMPER'])
+        cmd = '{}'.format(self.device.DEVICE_TOOLS['KEYCHAIN_DUMP'])
         msg = "Dumping the keychain"
         if self.options['filter']:
-            cmd += ' | grep -i "{}" -A 3 -B2'.format(self.options['filter'])
+            grep_args = ' -i -a "{}" -C 10 '.format(self.options['filter'])
```
Comment (marco-lancini) [inline needle/modules/storage/data/keychain_dump.py]: isn't this redundant now?

File: needle/modules/storage/data/keychain_dump.py (hunk 1 above)
Comment (marco-lancini) [inline needle/modules/storage/data/keychain_dump.py]: this creates plist files in the user's home directory. Is there any way to force a specified output directory?

File: needle/modules/storage/data/keychain_dump.py [hunk 3]
```diff
@@ -32,23 +35,54 @@ def module_pre(self):
...
+
+            # If this file exists
+            if self.device.remote_op.file_exist(filename):
+
+
+                files_exist = True
+
+                # read in the file either with or without the filter
+                if self.options['filter']:
+                    text = self.device.remote_op.read_file(filename,grep_args)
```
Comment (marco-lancini) [inline needle/modules/storage/data/keychain_dump.py]: I pushed a new version of `read_file`: if you don't have anything in `grep_args` you should pass `None`.
However I suggest to directly assign `grep_args = None` in the if statement of line 40

File: needle/modules/storage/data/keychain_dump.py [hunk 4]
```diff
@@ -32,23 +35,54 @@ def module_pre(self):
...
+                else:
+                    text = self.device.remote_op.read_file(filename)
+
+                # If the file (or filtered file) is not empty
+                if text != []:
```
Comment (marco-lancini) [inline needle/modules/storage/data/keychain_dump.py]: not clear why are you doing this. Can't you pull all plist files, and then analyze them locally?

File: needle/modules/storage/data/keychain_dump.py (hunk 1 above)
Comment (tghosth) [inline needle/modules/storage/data/keychain_dump.py]: Do you mean the user's home directory on the device? 

If so, it doesn't look like it is possible to pass a path to keychain_dump. Is there a built-in way in needle of changing the working directory on the device other than sending mkdir blah and then cd blah to exec_command?

File: needle/modules/storage/data/keychain_dump.py (hunk 2 above)
Comment (tghosth) [inline needle/modules/storage/data/keychain_dump.py]: Why is it redundant? Do you not still want them to have the filter option?

File: needle/modules/storage/data/keychain_dump.py (hunk 3 above)
Comment (tghosth) [inline needle/modules/storage/data/keychain_dump.py]: I will take a look.

File: needle/modules/storage/data/keychain_dump.py (hunk 4 above)
Comment (tghosth) [inline needle/modules/storage/data/keychain_dump.py]: You think I should pull the whole file to the local machine and then do the filter? That's not how the original worked.

**Comment by tghosth** [comment]
hi @marco-lancini, please see my comments above.

File: needle/modules/storage/data/keychain_dump.py (hunk 2 above)
Comment (marco-lancini) [inline needle/modules/storage/data/keychain_dump.py]: yes, but you are applying the filter later on (line 64). 
Keychain_dump create files, there is no need to grep directly when invoking it.

File: needle/modules/storage/data/keychain_dump.py (hunk 4 above)
Comment (marco-lancini) [inline needle/modules/storage/data/keychain_dump.py]: but originally the dump wasn't stored in 4 different plist files on the device's disk. If now we have to deal with those files, we might just retrieve them all (so the user can still analyze them manually later if he wants to)

File: needle/modules/storage/data/keychain_dump.py (hunk 2 above)
Comment (tghosth) [inline needle/modules/storage/data/keychain_dump.py]: ah ok, I understand

File: needle/modules/storage/data/keychain_dump.py (hunk 4 above)
Comment (tghosth) [inline needle/modules/storage/data/keychain_dump.py]: I understand. I will take a look.

**Comment by tghosth** [comment]
Second attempt in #46

**Pull request merged**